### PR TITLE
Add NWaveform Test

### PR DESCRIPTION
**Pull request by danestull**
### What does this Pull Request accomplish?

This change adds a system spot test that uses one of the NWaveform functions using the new sizing mechanism to ensure it still functions as appropriately

### Why should this Pull Request be merged?

This ensure we have a system test that checks the newly added functionality and that the responses we get out of it are appropriate. 

### What testing has been done?

New System test runs and passes with values that we expect the same as the C API example.
Screen Shot 2022-02-08 at 9 37 11 AM

### CODE REVIEW

File: source/tests/system/waveform_helpers.h [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+  auto c = TComplex{};
+  c.set_real(real);
+  c.set_imaginary(imaginary);
+  return c;
+}
+
+template <typename TFloat, typename TComplex>
+std::vector<TComplex> complex_array(
+    std::vector<TFloat> reals,
+    std::vector<TFloat> imaginaries)
```
Comment (astarche) [inline source/tests/system/waveform_helpers.h:40]: You probably got this from me, but these should be `const std::vector&`s.


```cpp
std::vector<TComplex> complex_array(
    const std::vector<TFloat>& reals,
    const std::vector<TFloat>& imaginaries)
```

File: source/tests/system/waveform_helpers.h [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+      reals.begin(),
+      reals.end(),
+      imaginaries.begin(),
+      std::back_inserter(c),
+      [](TFloat real, TFloat imaginary) { return complex<TFloat, TComplex>(real, imaginary); });
+  return c;
+}
+
+template <typename TFloat>
+TestIQData<TFloat> fetch_iq_data(nlohmann::json json)
```
Comment (astarche) [inline source/tests/system/waveform_helpers.h:54]: **fetch** is strongly associated with reading data from the driver in rfmx.

Does it work to just call this `load_test_iq_data` and have it be an overload of the file name version? Or `load_xyz_from_json`.?

File: source/tests/system/nirfmxwlan_driver_api_tests.cpp [hunk 3]
```diff
@@ -1152,6 +1163,54 @@ TEST_F(NiRFmxWLANDriverApiTests, TXPMIMOFromExample_FetchData_DataLooksReasonabl
   EXPECT_LT(0.0, txp_fetch_power_trace_response.power(0));
 }
 
+TEST_F(NiRFmxWLANDriverApiTests, AnalyzeNWaveformsIQ_FetchData_DataLooksReasonable)
+{
+  const auto num_frequency_segments = 1;
+  const auto num_receive_chains = 2;
+  double centerFrequency[2] = {5.18e9, 5.26e9};
+  std::vector<float64> IQx0{0, 0};
```
Comment (astarche) [inline source/tests/system/nirfmxwlan_driver_api_tests.cpp:1171]: nit: I try not to let `IQ` blow up my naming too much. `iq_x0`, `iq_dx`, `iq_size`, `iq` would work too.

File: source/tests/system/nirfmxwlan_driver_api_tests.cpp [hunk 4]
```diff
@@ -1152,6 +1163,54 @@ TEST_F(NiRFmxWLANDriverApiTests, TXPMIMOFromExample_FetchData_DataLooksReasonabl
   EXPECT_LT(0.0, txp_fetch_power_trace_response.power(0));
 }
 
+TEST_F(NiRFmxWLANDriverApiTests, AnalyzeNWaveformsIQ_FetchData_DataLooksReasonable)
+{
+  const auto num_frequency_segments = 1;
+  const auto num_receive_chains = 2;
+  double centerFrequency[2] = {5.18e9, 5.26e9};
```
Comment (reckenro) [inline source/tests/system/nirfmxwlan_driver_api_tests.cpp:1170]: Is it accurate to use `num_recieve_chains` here instead of 2?

**Review by reckenro — APPROVED** [review]

File: source/tests/system/waveform_helpers.h (hunk 1 above)
Comment (danestull) [reply source/tests/system/waveform_helpers.h]: updated

File: source/tests/system/nirfmxwlan_driver_api_tests.cpp [hunk 5]
```diff
@@ -1152,6 +1163,54 @@ TEST_F(NiRFmxWLANDriverApiTests, TXPMIMOFromExample_FetchData_DataLooksReasonabl
...
+  EXPECT_SUCCESS(session, client::ofdm_mod_acc_cfg_frequency_error_estimation_method(stub(), session, "", OFDM_MODACC_FREQUENCY_ERROR_ESTIMATION_METHOD_PREAMBLE_AND_PILOTS));
+  EXPECT_SUCCESS(session, client::ofdm_mod_acc_cfg_amplitude_tracking_enabled(stub(), session, "", OFDM_MODACC_AMPLITUDE_TRACKING_ENABLED_FALSE));
+  EXPECT_SUCCESS(session, client::ofdm_mod_acc_cfg_phase_tracking_enabled(stub(), session, "", OFDM_MODACC_PHASE_TRACKING_ENABLED_TRUE));
+  EXPECT_SUCCESS(session, client::ofdm_mod_acc_cfg_symbol_clock_error_correction_enabled(stub(), session, "", OFDM_MODACC_SYMBOL_CLOCK_ERROR_CORRECTION_ENABLED_TRUE));
+  EXPECT_SUCCESS(session, client::ofdm_mod_acc_cfg_channel_estimation_type(stub(), session, "", OFDM_MODACC_CHANNEL_ESTIMATION_TYPE_REFERENCE));
+
+  //READ TDMS File
+  auto waveforms = load_test_multiple_waveforms_data<float, nidevice_grpc::NIComplexNumberF32>("WLAN_80211n_20MHz_1Seg_2Chain_MIMO.json", 2);
+
+  for (int i = 0; i < 2; i++) {
```
[the remark this answers is not in the document]
Comment (danestull) [reply source/tests/system/nirfmxwlan_driver_api_tests.cpp]: added a comment for clarity

File: source/tests/system/waveform_helpers.h (hunk 2 above)
Comment (danestull) [reply source/tests/system/waveform_helpers.h]: updated this name to `load_iq_data_from_json`

File: source/tests/system/nirfmxwlan_driver_api_tests.cpp (hunk 3 above)
Comment (danestull) [reply source/tests/system/nirfmxwlan_driver_api_tests.cpp]: updated

File: source/tests/system/nirfmxwlan_driver_api_tests.cpp [hunk 6]
```diff
@@ -890,6 +900,7 @@ TEST_F(NiRFmxWLANDriverApiTests, DISABLED_OFDMModAccWithEVMBasedAutoLevelFromExa
   EXPECT_SUCCESS(session, client::ofdm_mod_acc_cfg_phase_tracking_enabled(stub(), session, "", OFDM_MODACC_PHASE_TRACKING_ENABLED_TRUE));
   EXPECT_SUCCESS(session, client::ofdm_mod_acc_cfg_symbol_clock_error_correction_enabled(stub(), session, "", OFDM_MODACC_SYMBOL_CLOCK_ERROR_CORRECTION_ENABLED_TRUE));
   EXPECT_SUCCESS(session, client::ofdm_mod_acc_cfg_channel_estimation_type(stub(), session, "", OFDM_MODACC_CHANNEL_ESTIMATION_TYPE_REFERENCE));
+
```
[the remark this answers is not in the document]
Comment (mgumble) [reply source/tests/system/nirfmxwlan_driver_api_tests.cpp:903]: Blank line still here unresolving conversation

File: source/tests/system/nirfmxwlan_driver_api_tests.cpp [hunk 7]
```diff
@@ -1152,6 +1163,53 @@ TEST_F(NiRFmxWLANDriverApiTests, TXPMIMOFromExample_FetchData_DataLooksReasonabl
...
+  EXPECT_SUCCESS(session, client::cfg_standard(stub(), session, "", STANDARD_802_11_N));
+  EXPECT_SUCCESS(session, client::cfg_channel_bandwidth(stub(), session, "", 20e6));
+  EXPECT_SUCCESS(session, client::select_measurements(stub(), session, "", MEASUREMENT_TYPES_OFDMMODACC, true));
+  EXPECT_SUCCESS(session, client::ofdm_mod_acc_cfg_measurement_length(stub(), session, "", 0, 16));
+  EXPECT_SUCCESS(session, client::ofdm_mod_acc_cfg_frequency_error_estimation_method(stub(), session, "", OFDM_MODACC_FREQUENCY_ERROR_ESTIMATION_METHOD_PREAMBLE_AND_PILOTS));
+  EXPECT_SUCCESS(session, client::ofdm_mod_acc_cfg_amplitude_tracking_enabled(stub(), session, "", OFDM_MODACC_AMPLITUDE_TRACKING_ENABLED_FALSE));
+  EXPECT_SUCCESS(session, client::ofdm_mod_acc_cfg_phase_tracking_enabled(stub(), session, "", OFDM_MODACC_PHASE_TRACKING_ENABLED_TRUE));
+  EXPECT_SUCCESS(session, client::ofdm_mod_acc_cfg_symbol_clock_error_correction_enabled(stub(), session, "", OFDM_MODACC_SYMBOL_CLOCK_ERROR_CORRECTION_ENABLED_TRUE));
+  EXPECT_SUCCESS(session, client::ofdm_mod_acc_cfg_channel_estimation_type(stub(), session, "", OFDM_MODACC_CHANNEL_ESTIMATION_TYPE_REFERENCE));
+  //READ TDMS File
```
Comment (mgumble) [inline source/tests/system/nirfmxwlan_driver_api_tests.cpp:1191]: nit: comments should start with a space.

File: source/tests/system/nirfmxwlan_driver_api_tests.cpp [hunk 8]
```diff
@@ -1152,6 +1163,53 @@ TEST_F(NiRFmxWLANDriverApiTests, TXPMIMOFromExample_FetchData_DataLooksReasonabl
...
+  auto waveforms = load_test_multiple_waveforms_data<float, nidevice_grpc::NIComplexNumberF32>("WLAN_80211n_20MHz_1Seg_2Chain_MIMO.json", 2);
+  //Concatinate waveform data, two dimension array passed as one dimension array with size array determining where one starts and one stops.
```
Comment (mgumble) [inline source/tests/system/nirfmxwlan_driver_api_tests.cpp:1193]: concatenate

**Review by mgumble — APPROVED** [review]

File: source/tests/system/nirfmxwlan_driver_api_tests.cpp (hunk 8 above)
Comment (danestull) [reply source/tests/system/nirfmxwlan_driver_api_tests.cpp]: fixed spelling

**Pull request merged**